pkg/scheduler/algorithm/priorities/scarce_resource_bin_packing.go

```diff
@@ -0,0 +1,126 @@
+	scarceResource string
+}
+
+// NewScarceResourceBinPacking creates a ScarceResourceBinPackingPriorityMap.
+func NewScarceResourceBinPacking(scarceResource string) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	scarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: scarceResource,
+	}
+	return scarceResourceBinPackingPrioritizer.ScarceResourceBinPackingPriorityMap, nil
+}
+
+// ScarceResourceBinPackingPriorityDefault creates a requestedToCapacity based
+// ResourceAllocationPriority using default resource scoring function shape.
+// The default function assigns 1.0 to resource when all capacity is available
+// and 0.0 when requested amount is equal to capacity.
+func ScarceResourceBinPackingPriorityDefault() *ScarceResourceBinPacking {
+	defaultscarceResourceBinPackingPrioritizer := &ScarceResourceBinPacking{
+		scarceResource: "",
+	}
+	return defaultscarceResourceBinPackingPrioritizer
+}
+
+// ScarceResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
+// It will detect whether the requested scarce resource is present on a node, and then calculate a score ranging from 0 to 10
+// based total utlization (best fit)
+// - If none of the scare resource are requested, this node will be given the lowest priority.
+// - If the scarce resource is requested, the larger the resource utlization ratio, the higher the node's priority.
+func (s *ScarceResourceBinPacking) ScarceResourceBinPackingPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
+	var score int
+	node := nodeInfo.Node()
+	if len(s.scarceResource) == 0 {
+		return schedulerapi.HostPriority{}, fmt.Errorf("scarceResource not defined")
+	}
+	if node == nil {
+		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
+	}
+	if !podRequestsResource(*pod, s.scarceResource) {
+		score = 0
+	} else {
+		score = calculateScareResourceScore(nodeInfo, pod.Spec.Containers, s.scarceResource)
+	}
+
+	return schedulerapi.HostPriority{
+		Host:  node.Name,
+		Score: score,
+	}, nil
+}
+
+// calculateScareResourceScore returns total utlization of the scare resource on the node
+func calculateScareResourceScore(nodeInfo *schedulercache.NodeInfo, containers []v1.Container, resource string) int {
+	reqResource := 0
+	usedResource := 0
+	for _, container := range containers {
+		if qunatity, ok := container.Resources.Requests[v1.ResourceName(resource)]; ok {
+			reqResource += int(qunatity.Value())
+		}
+	}
+	for _, pod := range nodeInfo.Pods() {
+		for _, container := range pod.Spec.Containers {
+			if qunatity, ok := container.Resources.Requests[v1.ResourceName(resource)]; ok {
+				usedResource += int(qunatity.Value())
+			}
+		}
+	}
+
+	available := int(nodeInfo.AllocatableResource().ScalarResources[v1.ResourceName(resource)])
+	return ((usedResource + reqResource) * schedulerapi.MaxPriority) / available
```

Comment:
why do we multiply this by maxpriority here ?

Reply:
Just trying to get the score in the scale from 1-10, so the user can control the priority effect by weight in the scheduler config. This can help when the user needs bin packing but also spread the pods evenly.